chandra_aca/centroid_resid.py

```diff
@@ -0,0 +1,325 @@
+    def zag_times(self, vals):
+        self._zag_times = np.array(vals)
+
+    def set_offsets(self):
+        """
+        Apply time offsets to centroids based on type and source of centroid, obsid
+        (suggesting 8x8 or 6x6 data), and aspect solution source.  These time offsets were
+        fit.  The offsets determined against OBC aspect solutions are quite variable and
+        this code will warn when applying offsets in the case when residuals are to be
+        calculated using an OBC solution.
+        """
+
+        if self._dt_applied is True:
+            raise ValueError("Attempted to apply centroid dt more than once")
+        is_or = self.obsid < 38000
+        # Offsets calculated using fit scripts in SKA/analysis/centroid_and_sol_time_offsets
+        # Also see plots in PR https://github.com/sot/chandra_aca/pull/25
+        if self.att_source == 'obc':
+            warnings.warn("Applying centroid time offset for {} centroids and obc att source.".format(
+                self.centroid_source))
+        # For ground centroids we don't need the fetch source
+        if self.centroid_source == "ground":
+            if is_or:
+                if self.att_source == 'ground':
+                    self.centroid_dt = 0.125
+                if self.att_source == 'obc':
+                    self.centroid_dt = 0.665
+            else:
+                if self.att_source == 'ground':
+                    self.centroid_dt = 0.054
+                if self.att_source == 'obc':
+                    self.centroid_dt = 0.087
+        else:
+            if len(fetch.data_source.sources()) > 1:
+                raise ValueError("Can't set offsets based on fetch data source if multiple data sources set")
+            fetch_source = fetch.data_source.sources()[0]
+            if fetch_source != 'cxc' and fetch_source != 'maude':
+                raise ValueError("Only maude and cxc fetch data sources are supported for offsets")
+            if is_or:
+                if fetch_source == 'cxc':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.4
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -1.8
+                if fetch_source == 'maude':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.8
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.34
+            else:
+                if fetch_source == 'cxc':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.44
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.43
+                elif fetch_source == 'maude':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.45
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.45
```

Comment:
All of this if/else nest looks disgusting to me, but I don't see particularly self-documenting ways out of it.  My original plan was to generalize the values close to -2.4, but that wasn't the suggestion in our meeting (which was instead to use the median values of each fit).  Also, even if I just use -2.4 for a bunch of these, the logic to get to the right cases would still be convoluted.  Should I use a dictionary of offsets with tuple keys or some such?